test/integration/README.md

```diff
@@ -0,0 +1,86 @@
+The Envoy integration test framework is designed to make it simple to test downstream-Envoy-upstream
+communication.  In the common case, one
+
+- Sends a request from downstream through Envoy
+- Verifies the request is received upstream, and possibly inspects elements of the headers or body
+- Sends a response from upstream through Envoy
+- Verifies the request is received downstream, again possibly inspecting headers and body.
+
+For the simplest variant of this, one could do the following.
+
+```
+// start Envoy, set up the fake upstreams.
+initialize();
+
+// Create a client aimed at Envoy’s default HTTP port.
+codec_client_ = makeHttpConnection(makeClientConnection((lookupPort("http"))));
+
+// Create some request headers.
+Http::TestHeaderMapImpl request_headers{{":method", "GET"},
+                                        {":path", "/test/long/url"},
+                                        {":scheme", "http"},
+                                        {":authority", "host"}};
+
+// Send the request headers from the client, wait until they are received upstream.  When they
+// are received, send the default response headers from upstream and wait until they are
+// received at by client
+sendRequestAndWaitForResponse(request_headers, 0, default_response_headers_, 0);
+
+// Verify the proxied request was received upstream, as expected.
+EXPECT_TRUE(upstream_request_->complete());
+EXPECT_EQ(0U, upstream_request_->bodyLength());
+// Verify the proxied response was received downstream, as expected.
+EXPECT_TRUE(response_->complete());
+EXPECT_STREQ("200", response_->headers().Status()->value().c_str());
+EXPECT_EQ(0U, response_->body().size());
+```
+
+Once you have the basic end to end test, it is fairly straight forward to modify it to test more
+interesting corner cases.   There are existing tests which send requests with bodies, have
```

Comment:
Nit: So many spaces between sentences, so much wow.

Reply:
Hahaha.  I actually had these fixed locally but didn't think it was worth its own push.  I could have spared you the horrors :-P